### PR TITLE
[G2M] CLI

### PR DESCRIPTION
**Pull request by woozyking**
basic CLI
- `togglProject` to replace `Projects` (no more split but to be lumped into one toggl project)
- `companyHolidays` support
- basic `process.argv` parsing, to be revised in the future with something like `yargs` or `minimist`

the new changes are reflected in generated 2021 timesheet

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -158,29 +159,42 @@ const genRange = ({
...
+          '2021-12-25',
+          '2021-12-26',
+          '2021-12-27',
+          '2021-12-28',
+          '2021-12-29',
+          '2021-12-30',
+          '2021-12-31',
+          '2022-01-01',
+          '2022-01-02',
+          '2022-01-03',
```
Comment (hyx131) [inline index.js:196]: _nitpick_: shall we include only weekdays and non statutory ones as `companyHolidays`? (half Dec 24, Dec 29-31, 2021)

**Review by hyx131 — APPROVED** [review]

File: index.js [hunk 2]
```diff
@@ -59,17 +56,15 @@ const genDaily = ({
...
-      // per-task, per-day error
-      const err = genError()
-      const d = moment.duration(hours * pportion * portion * err, 'hours')
-      durations[`${Project}::${Task}`] = {
-        startTime: startTime.format('HH:mm:ss'),
-        Duration: moment.utc(d.asMilliseconds()).format('HH:mm:ss'),
-      }
-      startTime.add(d)
-    })
+  shuffle(Object.entries(tasks)).filter(([, portion]) => portion > 0).forEach(([Task, portion]) => {
```
Comment (vxsl) [inline index.js:59]: what purpose does the lodash shuffle serve here? Seems like the randomness comes from the `genError` invocation below.

File: peepo.js [hunk 3]
```diff
@@ -85,5 +85,6 @@ module.exports.getVacays = async ({
 }
 
 if (require.main === module) {
-  this.getVacays({ after: '2020-01-01', before: '2020-12-31' }).then(JSON.stringify).then(console.log)
+  const args = process.argv.slice(2)
+  this.getVacays({ after: args[0], before: args[1] }).then(JSON.stringify).then(console.log)
```
Comment (vxsl) [inline peepo.js:89]: here too:

```suggestion
  const [after, before] = process.argv.slice(2)
  if (!after || !before) {
    throw new Error('Usage: [...]')
  }
  this.getVacays({ after, before }).then(JSON.stringify).then(console.log)
```

File: index.js (hunk 1 above)
Comment (woozyking) [reply index.js:196]: lol sure, this whole portion was auto-inserted by co-pilot as this is meant to be a one-off hardcoded hack, so didn't manually trim (except for the half Dec 24 handling, see below)

For the weekend we anyway have the logic in `genRange`

```js
// skip weekends and public holidays
if (HD.isHoliday(day.toDate()) || [6, 7].includes(day.isoWeekday())) {
  continue
}
```

for the half-day handling I did a quick Python pandas script to cap it at UTC 17:00 on that day for anyone's "work" extended beyond that point

File: index.js [hunk 4]
```diff
@@ -158,29 +159,42 @@ const genRange = ({
...
     })
   }
 }
 
 if (require.main === module) {
-  // TODO: parameterize these
-  const after = '2020-10-01' // inclusive
-  const before = '2020-12-31' // inclusive
-  const peepoSheet = './employees_2020.csv'
+  const [after, before, peepoSheet, togglProject = 'Engineering'] = process.argv.slice(2)
```
Comment (vxsl) [inline index.js:170]: Could do some validation here:
```suggestion
  const [after, before, peepoSheet, togglProject = 'Engineering'] = process.argv.slice(2)
  if (!after || !before || !peepoSheet || !togglProject) {
    throw new Error('Usage: [...]')
  }
```

**Review by vxsl — APPROVED** [review]

File: index.js (hunk 2 above)
Comment (woozyking) [reply index.js:59]: shuffle the order of the tasks

it'd be weird to see everyone repeats the tasks in the same order every day:
1. Development
2. Research/Design
3. QA/Maintenance
4. Admin/Ops

File: index.js (hunk 4 above)
Comment (woozyking) [reply index.js:170]: i'd defer this to when i iterate to use a more robust (and with auto-help/usage hinting) library

> basic process.argv parsing, to be revised in the future with something like yargs or minimist

Comment (vxsl) [reply index.js:170]: oh, right, sorry I missed that.

File: index.js (hunk 2 above)
Comment (vxsl) [reply index.js:59]: Oh I understand

**Pull request merged**